pkg/metering/writer.go

```diff
@@ -0,0 +1,231 @@
+		if len(c.LocalFS.BasePath) == 0 {
+			return errors.New("base path is required for the metering config")
+		}
+	default:
+		return errors.New("unsupported provider type for the metering config")
+	}
+	return nil
+}
+
+// Start starts the metering writer background loop
+func (mw *Writer) Start() {
+	if mw == nil {
+		return
+	}
+	mw.wg.Add(1)
+	go mw.meteringLoop()
+}
+
+// Stop stops the metering writer
+func (mw *Writer) Stop() {
+	if mw == nil {
+		return
+	}
+	if mw.cancel != nil {
+		mw.cancel()
+	}
+	mw.wg.Wait()
+	if mw.inner != nil {
+		mw.inner.Close()
+	}
+	log.Info("metering writer stopped")
+}
+
+// RegisterCollector registers a collector for a category. Overwrites existing if any.
+func (mw *Writer) RegisterCollector(collector Collector) {
+	if mw == nil {
+		return
+	}
+	mw.Lock()
+	defer mw.Unlock()
+	if mw.collectors == nil {
+		mw.collectors = make(map[string]Collector)
+	}
+	mw.collectors[collector.Category()] = collector
+}
+
+// meteringLoop runs the background loop to flush metering data periodically.
+func (mw *Writer) meteringLoop() {
+	defer logutil.LogPanic()
+	defer mw.wg.Done()
+
+	now := time.Now()
+	// Truncate to the nearest minute and add the interval to get the next flush time.
+	// By default, metering data should be flushed at the minute level with the correct timestamp.
+	// To ensure accuracy and avoid unexpected overwrites from overlapping timestamps, we need to round the time accordingly.
+	next := now.Truncate(flushInterval).Add(flushInterval)
+	log.Info("metering writer loop started",
+		zap.String("self-id", mw.id),
+		zap.Time("now", now),
+		zap.Time("next", next))
+	for {
+		select {
+		case <-mw.ctx.Done():
+			ts := next.Unix()
+			log.Info("context done received, flushing metering data one last time",
+				zap.Int64("timestamp", ts))
+			// Due to the context is already done, we need to use a new context to flush the data.
+			mw.flushMeteringData(context.Background(), ts)
+			log.Info("metering writer loop exits")
+			return
+		case <-time.After(next.Sub(now)):
+			mw.flushMeteringData(mw.ctx, next.Unix())
+			now = time.Now()
+			next = next.Add(flushInterval)
+		}
+	}
+}
+
+// flushMeteringData flushes aggregated metering data to the underlying storage
+func (mw *Writer) flushMeteringData(ctx context.Context, ts int64) {
+	collectors := mw.getCollectors()
+	if len(collectors) == 0 {
+		return
+	}
+
+	for category, collector := range collectors {
+		records := collector.Aggregate()
+		if len(records) == 0 {
+			continue
+		}
+		// This should never happen in the real world, it's used to avoid panic during testing.
+		if mw.inner == nil {
+			continue
+		}
+		meteringData := &common.MeteringData{
+			SelfID:    mw.id,
+			Timestamp: ts,
+			Category:  category,
+			Data:      records,
+		}
+		recordCount := len(records)
+		start := time.Now()
+		ctx, cancel := context.WithTimeout(ctx, flushTimeout)
+		// TODO: write with pagination if needed.
+		err := mw.inner.Write(ctx, meteringData)
```

Comment:
Shall we backoff some time (maybe in another routine) instead of directly return? Or is the write already implemented backoff logic?